src/main/java/org/ehcache/integrations/shiro/EhCache.java

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ehcache.integrations.shiro;
+
+import org.apache.shiro.cache.Cache;
+import org.apache.shiro.cache.CacheException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+public class EhCache<K, V> implements Cache<K, V> {
+
+    private static final Logger log = LoggerFactory.getLogger(EhCache.class);
+
+    private final org.ehcache.Cache<K, V> cache;
+
+    public EhCache(org.ehcache.Cache cache) {
+        if (cache == null) {
+            throw new IllegalArgumentException("Cache argument cannot be null.");
+        }
+
+        this.cache = cache;
+    }
+
+    public V get(K k) throws CacheException {
+        if (log.isTraceEnabled()) {
+            trace("Getting object", k);
+        }
+
+        if (k == null) {
+            return null;
+        }
+
+        V value = cache.get(k);
+        if (value == null) {
+            if (log.isTraceEnabled()) {
+                log.trace("Element for [" + k + "] is null.");
+            }
+        }
+
+        return value;
+    }
+
+    public V put(K k, V v) throws CacheException {
+        if (log.isTraceEnabled()) {
+            trace("Putting object", k);
+        }
+
+        V previousValue = get(k);
+        cache.put(k, v);
```

Comment:
Damn, it is yes.  Regardless I assume this is intended to be atomic... so we'll have to achieve this somehow.  Simplest solution would seem to be a loop involving `putIfAbsent(K, V)` and `replace(K, V)`.